# ¡Hola, Andrea!  

Mi nombre es Francisco Benavides, soy code reviewer de Tripleten y voy a revisar el proyecto que acabas de desarrollar.


Cada ves que vea un error por primera vez, lo señalaré. Deberás encontrarlo y arreglarlo, ya que buscamos que te prepares para un trabajo real, en el cual tu líder de equipo hará lo mismo. Si no puedes solucionar el error, te brindaré algunos tips más especificos la proxima vez.

Podrás encontrar mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer cuidadosamente cada una de tus implementaciones que has llevado a cabo para complir con tu informe. Cada uno de mis comentarios tiene el siguiente código de colores:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Si encuentras que es necesario, puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

Mucho éxito!

# Proyecto integrado 10

El objetivo del presente proyecto es encontrar los mejores lugares para abrir 200 pozos nuevos de petróleo. 

# Contenido <a id='back'></a>

* [Etapa 1. Abrir el arhivo de datos y análisis general](#data_review)
* [Etapa 2. Entrenamiento y evaluación del modelo para cada región](#train_test)
* [Etapa 3. Cálculo de ganancias](#profit)
* [Etapa 4. Bootstrapping](#bootstrapping)
* [Etapa 5. Conclusión](#end)

## Etapa 1. Abrir el archivo de datos y análisis general <a id='data_review'></a>

In [12]:
#Llamar a las librerías
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

In [13]:
#Llamar a las bases de datos
data_region_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_region_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_region_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [14]:
#Observar la información general de las datas
print(data_region_0.info())
print()
print(data_region_1.info())
print()
print(data_region_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column

In [15]:
print(data_region_0.head())
print()
print(data_region_1.head())
print()
print(data_region_2.head())

      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


In [16]:
print(data_region_0.describe())
print()
print(data_region_1.describe())
print()
print(data_region_2.describe())

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%      

In [17]:
#Ver si hay duplicados
print(data_region_0.duplicated().sum())
print(data_region_1.duplicated().sum())
print(data_region_2.duplicated().sum())

0
0
0


### Conclusión etapa 1
Puede observarse que las variables están categorizadas de manera correcta y que no hay valores perdidos ni valores duplicados. Además, se observa que la desviación estándar del volumen de reserva en los pozos de petróleo por región es alta y muy similar entre las tres. En ese contexto, es necesario mencionar que la región con la mediana más alta es la región 2, teniendo la mitad de sus pozos un volumen de reservas por encima de los 94.93 mil barriles. Le sigue la región 0, teniendo la mitad de sus pozos un volumen de reservas por encima de los 91.85 mil barriles. Finalmente, se encuentra la región 1, teniendo la mitad de sus pozos un volumen de reservas por encima de los 57.09 mil barriles.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen analisis de la etapa 1
</div>

## Etapa 2. Entrenamiento y evaluación del modelo para cada región <a id='train_test'></a>

In [18]:
#Separación de las características y el target para cada región
features_0 = data_region_0.drop(columns=['id','product']).select_dtypes(include=[np.number])
target_0 = data_region_0['product']

features_1 = data_region_1.drop(columns=['id','product']).select_dtypes(include=[np.number])
target_1 = data_region_1['product']

features_2 = data_region_2.drop(columns=['id','product']).select_dtypes(include=[np.number])
target_2 = data_region_2['product']

#División en conjuntos de entrenamiento y validación
features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(features_0, target_0, test_size=0.25, random_state=12345)
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(features_1, target_1, test_size=0.25, random_state=12345)
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(features_2, target_2, test_size=0.25, random_state=12345)
    
#Entrenamiento del modelo de regresión lineal
model_region_0 = LinearRegression()
model_region_0.fit(features_train_0, target_train_0)

model_region_1 = LinearRegression()
model_region_1.fit(features_train_1, target_train_1)

model_region_2 = LinearRegression()
model_region_2.fit(features_train_2, target_train_2)
    
#Predicciones en el conjunto de validación
predictions_region_0 = model_region_0.predict(features_valid_0)
predictions_region_1 = model_region_1.predict(features_valid_1)
predictions_region_2 = model_region_2.predict(features_valid_2)
    
#Cálculo del RMSE
rmse_region_0 = np.sqrt(mean_squared_error(target_valid_0, predictions_region_0))
rmse_region_1 = np.sqrt(mean_squared_error(target_valid_1, predictions_region_1))
rmse_region_2 = np.sqrt(mean_squared_error(target_valid_2, predictions_region_2))

print(f"RMSE Región 0: {rmse_region_0}")
print(f"RMSE Región 1: {rmse_region_1}")
print(f"RMSE Región 2: {rmse_region_2}")
    
#Volumen promedio de reservas predicho
predicted_mean_region_0 = predictions_region_0.mean()
predicted_mean_region_1 = predictions_region_1.mean()
predicted_mean_region_2 = predictions_region_2.mean()

RMSE Región 0: 37.5794217150813
RMSE Región 1: 0.893099286775617
RMSE Región 2: 40.02970873393434


In [19]:
predicted_mean_region_0 = predictions_region_0.mean()
predicted_mean_region_1 = predictions_region_1.mean()
predicted_mean_region_2 = predictions_region_2.mean()

print(f"Volumen promedio predicho para la Región 0: {predicted_mean_region_0:.2f}")
print(f"Volumen promedio predicho para la Región 1: {predicted_mean_region_1:.2f}")
print(f"Volumen promedio predicho para la Región 2: {predicted_mean_region_2:.2f}")

Volumen promedio predicho para la Región 0: 92.59
Volumen promedio predicho para la Región 1: 68.73
Volumen promedio predicho para la Región 2: 94.97


### Conclusión etapa 2
Se observa que el modelo con mejor ajuste es el de la región 1, debido a que el RMSE es de 0.89. Lo anterior indica que este modelo predice muy bien el volumen de reservas en esta región. En contraste, el RMSE de las regiones 0 y 2 es considerablemente más alto. Adicionalmente, es posible observar que en lo que respecta al volumen de los pozos de petróleo predichos, la región 2 es la que ostenta mayor cantidad, con 94.97 miles de barriles. Le sigue la región 0 con 92.59 miles de barriles y, en tercer lugar, estaría la región 1, con 68.73 miles de barriles. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo entrenando cada modelo de las regiones
</div>

## Etapa 3. Cálculo de ganancias <a id='profit'></a>

In [20]:
#Valores clave
budget = 100000000  #Presupuesto en dólares
best_points = 200
revenue_per_unit = 4500  #Ingreso por unidad de producto

#Umbral mínimo de reservas para evitar pérdidas
umbral = budget/(best_points*revenue_per_unit)
print(f"Umbral mínimo de reservas para evitar pérdidas: {umbral:.2f}")

Umbral mínimo de reservas para evitar pérdidas: 111.11


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
**solucionado**    
    
    
El umbral de reservas se calcula como:
    
    budget = 100000000
    best_points = 200
    revenue_per_unit = 4500
    
    umbral = budget/(best_points*revenue_per_unit)
    
</div>


In [21]:
#Función para el cálculo de las ganancias
def calculate_profit(predictions, target, count=200):
    temp_df = pd.DataFrame({'predictions': predictions, 'target': target})
    
    #Mejores 200 pozos
    top_predictions = temp_df.nlargest(count, 'predictions')
    
    #Cálculo de las ganancias usando los valores reales
    revenue = top_predictions['target'].sum() * revenue_per_unit
    profit = revenue - budget
    return profit

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
**solucionado**    
    
Los 200 mejores pozos debes elegirlos con las predicciones, pero el revenue debes calcularlo con los valores reales de los pozos
</div>

In [22]:
#Cálculo de ganancias para cada región
profit_region_0 = calculate_profit(predictions_region_0, target_valid_0)
profit_region_1 = calculate_profit(predictions_region_1, target_valid_1)
profit_region_2 = calculate_profit(predictions_region_2, target_valid_2)

print(f'Ganancia estimada para la Región 0: ${profit_region_0:.2f}')
print(f'Ganancia estimada para la Región 1: ${profit_region_1:.2f}')
print(f'Ganancia estimada para la Región 2: ${profit_region_2:.2f}')

Ganancia estimada para la Región 0: $33208260.43
Ganancia estimada para la Región 1: $24150866.97
Ganancia estimada para la Región 2: $27103499.64


### Conclusión etapa 3
Considerando que el volumen mínimo de reservas qe un pozo debe tener para ser rentable es de 111.11 mil barriles, así como los mejores 200 pozos de petróleo por región, fue posible determinar las ganacias estimdas por región. Estas ganancias nos indican que en la región 0 se obtendrían más ganancias a comparación de las otras dos regiones, superando las ganancias de la región 2 en 6 millones y de la región 1 en 9 millones. Sin embargo, debe mencionarse que ninguna de las 3 regiones analizadas cumple con el umbral mínimo de reservas, significando ello que existe un riesgo inherente en las 3 regiones.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
buen trabajo calculando los valores base de cada pozo
</div>

## Etapa 4. Bootstrapping <a id='bootstrapping'></a>

In [38]:
#Función para evaluación con bootstrapping
def bootstrapped_profit(predictions, target, count=200, n_bootstrap=1000, sample_size=500):
    temp_df = pd.DataFrame({'predictions': predictions, 'target': target})

    profits = []

    for i in range(n_bootstrap):
        #Muestreo aleatorio con reemplazo
        sample = temp_df.sample(n=sample_size, replace=True, random_state=12345+i)
        top_predictions = sample.nlargest(count, 'predictions')


        #Cálculo de la ganancia
        revenue = top_predictions['target'].sum() * revenue_per_unit
        profit = revenue - budget
        profits.append(profit)

    profits = pd.Series(profits)
    lower_bound = profits.quantile(0.025)
    upper_bound = profits.quantile(0.975)
    mean_profit = profits.mean()
    risk = (profits < 0).sum()/len(profits)

    return mean_profit, lower_bound, upper_bound, risk

In [39]:
#Aplicar bootstrapping a cada región
mean_profit_0, lower_0, upper_0, risk_0 = bootstrapped_profit(predictions_region_0, target_valid_0)
mean_profit_1, lower_1, upper_1, risk_1 = bootstrapped_profit(predictions_region_1, target_valid_1)
mean_profit_2, lower_2, upper_2, risk_2 = bootstrapped_profit(predictions_region_2, target_valid_2)

print(f'Región 0 - Ganancia promedio: ${mean_profit_0:.2f}, Intervalo de confianza: (${lower_0:.2f}, ${upper_0:.2f}), Riesgo: {risk_0:.2%}')
print(f'Región 1 - Ganancia promedio: ${mean_profit_1:.2f}, Intervalo de confianza: (${lower_1:.2f}, ${upper_1:.2f}), Riesgo: {risk_1:.2%}')
print(f'Región 2 - Ganancia promedio: ${mean_profit_2:.2f}, Intervalo de confianza: (${lower_2:.2f}, ${upper_2:.2f}), Riesgo: {risk_2:.2%}')


Región 0 - Ganancia promedio: $3947268.35, Intervalo de confianza: ($-1418370.01, $8952001.68), Riesgo: 6.90%
Región 1 - Ganancia promedio: $4468417.70, Intervalo de confianza: ($423838.71, $8608745.86), Riesgo: 1.20%
Región 2 - Ganancia promedio: $3855086.50, Intervalo de confianza: ($-1738854.19, $9272058.12), Riesgo: 8.10%


### Conclusión etapa 4
En primer lugar, se debe mencionar que la región con el menor riesgo es la región 1 (1.20%), seguido de la región 0 (6.90%) y de la región 2 (8.10%). Además, debe mencionarse que la región con la mayor ganancia promedio es la región 1, siendo esta aprox 500 mil más que la región con las segundas mejores ganancias (la región 0). Adicionalmente, debe mencionarse que tanto la región 0, como la 2, tienen valores negativos, lo que implicaría que hay un riesgo de que puedan existir pérdidas. Todo lo anterior indica que la región 1 sería la mejor opción desde una perspectiva de rentabilidad y riesgo. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
  
**solucionado**    
    
**3° iteración.**
 
Sigues manteniendo el error en el boostrap, debes elegir la muestra de boostrap desde todos los valores, no solo los mejores 200. Luego una vez obtenida la muestra de boostrap elegir los mejores 200 de esa muestra y calcular el revenue
  
Recuerda que primero debes generar la muestra de boostrap y luego elegir los 200 mejores pozos, tu codigo quedaria asi:
    
    def bootstrapped_profit(predictions, target, count=200, n_bootstrap=1000, sample_size=500):
    temp_df = pd.DataFrame({'predictions': predictions, 'target': target})

    profits = []

    #Selección de los 200 mejores pozos
    top_predictions = temp_df.nlargest(count, 'predictions')
    
    for i in range(n_bootstrap):
        #Muestreo aleatorio con reemplazo
        sample = temp_df.sample(n=sample_size, replace=True, random_state=12345+i)
        top_predictions = sample.nlargest(count, 'predictions')


        #Cálculo de la ganancia
        revenue = top_predictions['target'].sum() * revenue_per_unit
        profit = revenue - budget
        profits.append(profit)

        profits = pd.Series(profits)
        lower_bound = profits.quantile(0.025)
        upper_bound = profits.quantile(0.975)
        mean_profit = profits.mean()
        risk = (profits < 0).sum()/len(profits)

        return mean_profit, lower_bound, upper_bound, risk

    
    
**2° iteracion**
    
Recuerda que en el enunciado mencionan que se estudian 500 pozos,por lo que las muestras de bootstrap deben ser de ese tamaño, no del largo de temp_df
    
**1° iteracion**
Gran trabajo!, como detalle El riesgo se calcula como (values < 0).sum()/len(values)
</div>

## Etapa 5. Conclusión <a id='end'></a>

Por lo expuesto con anteriordad, se concluye que la mejor región para abrir 200 pozos de petróleo es la región 1. Lo anterior se debe a que esta región es la más segura y predecible. Además, el modelo tiene alta precisión y el riesgo muy bajo, lo que la hace una región estable y una inversión más segura.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen trabajo y ordenado!
</div>